sslmodule_init_exceptions(PyObject *module)
{
    PyObject *bases = NULL;

#define add_exception(exc, name, doc, base)                               \
do {                                                                      \
    (exc) = PyErr_NewExceptionWithDoc("ssl." name, (doc), (base), NULL);  \
    if ((exc) == NULL) goto error;                                        \
    if (PyModule_AddObjectRef(module, name, exc) < 0) goto error;         \
} while(0)

    PySSLErrorObject = PyType_FromSpecWithBases(&sslerror_type_spec, PyExc_OSError);
    if (PySSLErrorObject == NULL) {
        goto error;
    }
    if (PyModule_AddObjectRef(module, "SSLError", PySSLErrorObject) < 0) {
        goto error;
    }

    /* ssl.CertificateError used to be a subclass of ValueError */
    bases = PyTuple_Pack(2, PySSLErrorObject, PyExc_ValueError);
    if (bases == NULL) {
        goto error;
    }
    add_exception(
        PySSLCertVerificationErrorObject,
        "SSLCertVerificationError",
        SSLCertVerificationError_doc,
        bases
    );
    Py_CLEAR(bases);

    add_exception(
        PySSLZeroReturnErrorObject,
        "SSLZeroReturnError",
        SSLZeroReturnError_doc,
        PySSLErrorObject
    );

    add_exception(
        PySSLWantWriteErrorObject,
        "SSLWantWriteError",
        SSLWantWriteError_doc,
        PySSLErrorObject
    );

    add_exception(
        PySSLWantReadErrorObject,
        "SSLWantReadError",
        SSLWantReadError_doc,
        PySSLErrorObject
    );

    add_exception(
        PySSLSyscallErrorObject,
        "SSLSyscallError",
        SSLSyscallError_doc,
        PySSLErrorObject
    );

    add_exception(
        PySSLEOFErrorObject,
        "SSLEOFError",
        SSLEOFError_doc,
        PySSLErrorObject
    );
#undef add_exception

    return 0;
  error:
    Py_XDECREF(bases);
    return -1;
}
